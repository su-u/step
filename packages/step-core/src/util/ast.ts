import { IInterpreterRules } from '../interpriterRules/types';

export const getIndex = ({ ast, execObject: { manager }}: IInterpreterRules<any>) => {
  if (ast.children.NumberLiteral) {
    return parseInt(ast.children.NumberLiteral[0].image);
  } else if (ast.children.Identifier) {
    return manager.variable.reference(ast.children.Identifier[0].image).image;
  }
};
