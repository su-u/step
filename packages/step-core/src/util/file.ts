import * as fs from 'fs-extra';

export const writeAstToJson = (ast: any): void => {
  fs.writeFile('out.json', JSON.stringify(ast, undefined, 2), (err) => {
    if (err) {
      console.error(err);
    } else {
      console.info('write end');
    }
  });
};
