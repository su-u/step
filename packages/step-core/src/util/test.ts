import { parseInput } from '../parser';
import { Manager } from '../manager';
import { interpreter } from '../interpreter';

export const sourceParseJson = (source: string) => {
  return JSON.stringify(parseInput(source), undefined, 2);
};

export const exec = (text: string, manager: Manager) => {
  const ast = parseInput(text);
  interpreter({ ast, context: { manager, interpreter } });
  return manager;
};
