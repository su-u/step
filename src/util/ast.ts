import { IInterpreterRules } from '../interpriterRules/types';

export const getIndex = ({ ast, manager }: IInterpreterRules) => {
  if (ast.children.NumberLiteral !== undefined) {
    return parseInt(ast.children.NumberLiteral[0].image);
  } else if (ast.children.Identifier !== undefined) {
    return manager.variable.reference(ast.children.Identifier[0].image).image;
  }
};
