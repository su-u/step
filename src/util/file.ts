import * as fs from 'fs-extra';
import logger from '@/logger';

export const writeAstToJson = (ast: AstType): void => {
  fs.writeFile('out.json', JSON.stringify(ast, undefined, 2), (err) => {
    if (err) {
      logger.error(err);
    } else {
      logger.info('write end');
    }
  });
};
