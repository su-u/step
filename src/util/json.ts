export const toJson = (ast: any) => {
  return JSON.stringify(ast, undefined, 2);
};
