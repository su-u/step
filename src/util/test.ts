import { parseInput } from '@/parser';

export const sourceParse = (source: string) => {
  return JSON.stringify(parseInput(source), undefined, 2);
};
