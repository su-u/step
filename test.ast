Ruleset = syntax.pegjs
source code = example4.chb
{
  type: 'Program',
  body: [
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'a' },
        right: { type: 'Literal', value: 'foo', class: 'String' }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'b' },
        right: { type: 'Literal', value: 'bar', class: 'String' }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'c' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Identifier', name: 'a' },
          right: { type: 'Identifier', name: 'b' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'd' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Literal', value: 1, class: 'Number' },
          right: { type: 'Literal', value: 5, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'e' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Identifier', name: 'c' },
          right: { type: 'Identifier', name: 'd' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'f' },
        right: {
          type: 'BinaryExpression',
          operator: '>',
          left: { type: 'Literal', value: 3, class: 'Number' },
          right: { type: 'Literal', value: 4, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'g' },
        right: {
          type: 'BinaryExpression',
          operator: '<',
          left: { type: 'Literal', value: 3, class: 'Number' },
          right: { type: 'Literal', value: 4, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'h' },
        right: {
          type: 'BinaryExpression',
          operator: '<=',
          left: {
            type: 'BinaryExpression',
            operator: '+',
            left: { type: 'Literal', value: 1, class: 'Number' },
            right: {
              type: 'BinaryExpression',
              operator: '*',
              left: { type: 'Literal', value: 2, class: 'Number' },
              right: { type: 'Literal', value: 3, class: 'Number' }
            }
          },
          right: { type: 'Literal', value: 5, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'i' },
        right: {
          type: 'BinaryExpression',
          operator: '=',
          left: {
            type: 'BinaryExpression',
            operator: '/',
            left: { type: 'Literal', value: 8, class: 'Number' },
            right: { type: 'Literal', value: 4, class: 'Number' }
          },
          right: {
            type: 'BinaryExpression',
            operator: '/',
            left: { type: 'Literal', value: 18, class: 'Number' },
            right: { type: 'Literal', value: 9, class: 'Number' }
          }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'j' },
        right: {
          type: 'BinaryExpression',
          operator: '=',
          left: { type: 'Literal', value: '6', class: 'String' },
          right: { type: 'Identifier', name: 'd' }
        }
      }
    }
  ]
}
