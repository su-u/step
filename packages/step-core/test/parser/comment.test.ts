import { sourceParseJson } from '../../src/util/test';

describe('コメント', () => {
  test('1', () => {
    const source = `
# コメント
## コメント
### コメント #
    `;
    expect(sourceParseJson(source)).toMatchSnapshot();
  });
});
