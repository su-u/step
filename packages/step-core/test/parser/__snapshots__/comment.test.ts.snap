// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`コメント 1 1`] = `
"{
  "name": "ProgramRoot",
  "children": {
    "rules": [
      {
        "name": "ProgramRules",
        "children": {
          "rules": [
            {
              "image": "# コメント\\n",
              "startOffset": 1,
              "endOffset": 7,
              "startLine": 2,
              "endLine": 2,
              "startColumn": 1,
              "endColumn": 7,
              "tokenTypeIdx": 3,
              "tokenType": {
                "name": "Comment",
                "PATTERN": {},
                "tokenTypeIdx": 3,
                "CATEGORIES": [],
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false
              }
            }
          ]
        }
      },
      {
        "name": "ProgramRules",
        "children": {
          "rules": [
            {
              "image": "## コメント\\n",
              "startOffset": 8,
              "endOffset": 15,
              "startLine": 3,
              "endLine": 3,
              "startColumn": 1,
              "endColumn": 8,
              "tokenTypeIdx": 3,
              "tokenType": {
                "name": "Comment",
                "PATTERN": {},
                "tokenTypeIdx": 3,
                "CATEGORIES": [],
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false
              }
            }
          ]
        }
      },
      {
        "name": "ProgramRules",
        "children": {
          "rules": [
            {
              "image": "### コメント #\\n",
              "startOffset": 16,
              "endOffset": 26,
              "startLine": 4,
              "endLine": 4,
              "startColumn": 1,
              "endColumn": 11,
              "tokenTypeIdx": 3,
              "tokenType": {
                "name": "Comment",
                "PATTERN": {},
                "tokenTypeIdx": 3,
                "CATEGORIES": [],
                "categoryMatches": [],
                "categoryMatchesMap": {},
                "isParent": false
              }
            }
          ]
        }
      }
    ]
  }
}"
`;
